File: src/services/automation.ts
/**
 * Automation Orchestrator
 *
 * Coordinates detection, search triggering, and logging for complete
 * automation cycles. Handles partial failures gracefully.
 */

import { detectAll } from "./detector";
import { triggerSearches } from "./search-trigger";
import {
  logCycleStart,
  logCycleEnd,
  logSearches,
  logServerError,
  logSearchError,
} from "../lib/logger";

/** Result of a complete automation cycle */
export interface CycleResult {
  success: boolean;
  detectionResults: {
    totalMissing: number;
    totalCutoff: number;
    successCount: number;
    failureCount: number;
  };
  searchResults: {
    missingTriggered: number;
    cutoffTriggered: number;
    successCount: number;
    failureCount: number;
  };
  totalSearches: number;
  totalFailures: number;
  errors: string[];
}

/**
 * Execute a complete automation cycle
 *
 * @param isManual - Whether this is a manual trigger (vs scheduled)
 * @param dryRun - If true, preview what would be searched without triggering actual searches
 * @returns Cycle result summary
 */
export async function runAutomationCycle(
  isManual = false,
  dryRun = false
): Promise<CycleResult> {
  const errors: string[] = [];

  // Log cycle start (only if not dry-run)
  if (!dryRun) {
    logCycleStart(isManual);
  }

  // Phase 1: Detection
  const detectionResults = await detectAll();

  // Log detection failures (only if not dry-run)
  for (const result of detectionResults.results) {
    if (result.error) {
      if (!dryRun) {
        logServerError(result.serverName, result.serverType, result.error);
      }
      errors.push(`Detection failed for ${result.serverName}: ${result.error}`);
    }
  }

  // Phase 2: Search Triggering (or preview in dry-run mode)
  let searchResults;
  try {
    searchResults = await triggerSearches(detectionResults, dryRun);
  } catch (error) {
    const errorMsg = error instanceof Error ? error.message : String(error);
    errors.push(`Search triggering failed: ${errorMsg}`);

    // Log cycle end with failure (only if not dry-run)
    if (!dryRun) {
      logCycleEnd(0, errors.length, isManual);
    }

    return {
      success: false,
      detectionResults: {
        totalMissing: detectionResults.totalMissing,
        totalCutoff: detectionResults.totalCutoff,
        successCount: detectionResults.successCount,
        failureCount: detectionResults.failureCount,
      },
      searchResults: {
        missingTriggered: 0,
        cutoffTriggered: 0,
        successCount: 0,
        failureCount: 0,
      },
      totalSearches: 0,
      totalFailures: errors.length,
      errors,
    };
  }

  // Log successful searches (only if not dry-run)
  if (!dryRun) {
    const serverSearchCounts = new Map<
      string,
      { missing: number; cutoff: number }
    >();

    for (const result of searchResults.results) {
      const key = `${result.serverName}:${result.serverType}`;

      if (!serverSearchCounts.has(key)) {
        serverSearchCounts.set(key, { missing: 0, cutoff: 0 });
      }

      const counts = serverSearchCounts.get(key)!;

      if (result.success) {
        if (result.category === "missing") {
          counts.missing += result.itemIds.length;
        } else {
          counts.cutoff += result.itemIds.length;
        }
      } else {
        // Log search failures
        logSearchError(
          result.serverName,
          result.serverType,
          result.category,
          result.error ?? "Unknown error"
        );
        errors.push(
          `Search trigger failed for ${result.serverName} (${result.category}): ${result.error}`
        );
      }
    }

    // Log aggregated successful searches per server
    for (const [key, counts] of serverSearchCounts) {
      const [serverName, serverType] = key.split(":");

      if (counts.missing > 0) {
        logSearches(
          serverName,
          serverType as "radarr" | "sonarr",
          "missing",
          counts.missing,
          isManual
        );
      }

      if (counts.cutoff > 0) {
        logSearches(
          serverName,
          serverType as "radarr" | "sonarr",
          "cutoff",
          counts.cutoff,
          isManual
        );
      }
    }
  }

  const totalSearches =
    searchResults.missingTriggered + searchResults.cutoffTriggered;
  const totalFailures =
    detectionResults.failureCount + searchResults.failureCount;

  // Log cycle end (only if not dry-run)
  if (!dryRun) {
    logCycleEnd(totalSearches, totalFailures, isManual);
  }

  return {
    success: totalFailures === 0,
    detectionResults: {
      totalMissing: detectionResults.totalMissing,
      totalCutoff: detectionResults.totalCutoff,
      successCount: detectionResults.successCount,
      failureCount: detectionResults.failureCount,
    },
    searchResults: {
      missingTriggered: searchResults.missingTriggered,
      cutoffTriggered: searchResults.cutoffTriggered,
      successCount: searchResults.successCount,
      failureCount: searchResults.failureCount,
    },
    totalSearches,
    totalFailures,
    errors,
  };
}

/**
 * Get a human-readable summary of cycle results
 */
export function formatCycleResult(result: CycleResult): string {
  const lines: string[] = [];

  lines.push("=== Automation Cycle Summary ===");
  lines.push("");

  // Detection summary
  lines.push("Detection:");
  lines.push(
    `  Found: ${result.detectionResults.totalMissing} missing, ${result.detectionResults.totalCutoff} cutoff`
  );
  lines.push(
    `  Servers: ${result.detectionResults.successCount} successful, ${result.detectionResults.failureCount} failed`
  );
  lines.push("");

  // Search summary
  lines.push("Search Triggering:");
  lines.push(
    `  Triggered: ${result.searchResults.missingTriggered} missing, ${result.searchResults.cutoffTriggered} cutoff`
  );
  lines.push(
    `  Total: ${result.totalSearches} searches, ${result.searchResults.failureCount} failures`
  );
  lines.push("");

  // Overall status
  if (result.success) {
    lines.push("Status: ✓ Success");
  } else {
    lines.push(`Status: ✗ ${result.totalFailures} failures`);
    if (result.errors.length > 0) {
      lines.push("");
      lines.push("Errors:");
      for (const error of result.errors) {
        lines.push(`  - ${error}`);
      }
    }
  }

  return lines.join("\n");
}

File: src/services/automation.go
package services

import (
	"context"
	"fmt"
	"time"

	"github.com/user/janitarr/src/database"
	"github.com/user/janitarr/src/logger"
)

// AutomationDetector defines the interface for content detection.
type AutomationDetector interface {
	DetectAll(ctx context.Context) (*DetectionResults, error)
}

// AutomationSearchTrigger defines the interface for triggering searches.
type AutomationSearchTrigger interface {
	TriggerSearches(ctx context.Context, detectionResults *DetectionResults, limits database.SearchLimits, dryRun bool) (*TriggerResults, error)
}

// AutomationLogger defines the interface for logging automation events.
type AutomationLogger interface {
	LogCycleStart(isManual bool) *logger.LogEntry
	LogCycleEnd(totalSearches, failures int, isManual bool) *logger.LogEntry
	LogDetectionComplete(serverName, serverType string, missing, cutoffUnmet int) *logger.LogEntry
	LogSearches(serverName, serverType, category string, count int, isManual bool) *logger.LogEntry
	LogServerError(serverName, serverType, reason string) *logger.LogEntry
	LogSearchError(serverName, serverType, category, reason string) *logger.LogEntry
}

// AutomationDB defines the interface for database operations needed by Automation.
// Note: AddLogEntry removed as logger handles that.
type AutomationDB interface {
	GetAppConfig() database.AppConfig
}

// Automation orchestrates the detection, search triggering, and logging process.
type Automation struct {
	db       AutomationDB
	detector AutomationDetector
	trigger  AutomationSearchTrigger
	logger   AutomationLogger
}

// NewAutomation creates a new Automation service.
// The db parameter should now represent the database operations needed by Automation,
// and the logger parameter should be a logger.Logger instance.
func NewAutomation(db AutomationDB, detector AutomationDetector, trigger AutomationSearchTrigger, appLogger AutomationLogger) *Automation {
	return &Automation{
		db:       db,
		detector: detector,
		trigger:  trigger,
		logger:   appLogger,
	}
}

// RunCycle executes a full automation cycle: detect, trigger searches, and log results.
func (a *Automation) RunCycle(ctx context.Context, isManual, dryRun bool) (*CycleResult, error) {
	startTime := time.Now()
	a.logger.LogCycleStart(isManual)

	cycleResult := &CycleResult{
		Success: true,
		Errors:  []string{},
	}

	// 1. Get application configuration for search limits
	config := a.db.GetAppConfig()

	// 2. Detect missing and cutoff content
	detectionResults, err := a.detector.DetectAll(ctx)
	if err != nil {
		cycleResult.Success = false
		cycleResult.Errors = append(cycleResult.Errors, fmt.Sprintf("detection failed: %v", err))
		// Continue with partial detection results if any were returned
	}
	cycleResult.DetectionResults = *detectionResults

	// Log detection completion and errors for each server
	for _, res := range detectionResults.Results {
		if res.Error != "" {
			if !dryRun { // Added condition for dryRun
				a.logger.LogServerError(res.ServerName, res.ServerType, fmt.Sprintf("detection error: %s", res.Error))
			}
			cycleResult.Success = false
			cycleResult.TotalFailures++
			cycleResult.Errors = append(cycleResult.Errors, fmt.Sprintf("server %s detection failed: %s", res.ServerName, res.Error))
		} else {
			// Log successful detection completion
			a.logger.LogDetectionComplete(res.ServerName, res.ServerType, len(res.Missing), len(res.Cutoff))
		}
	}
	// 3. Trigger searches
	triggerResults, err := a.trigger.TriggerSearches(ctx, detectionResults, config.SearchLimits, dryRun)
	if err != nil {
		cycleResult.Success = false
		cycleResult.Errors = append(cycleResult.Errors, fmt.Sprintf("triggering searches failed: %v", err))
		// Continue with partial trigger results if any were returned
	}
	cycleResult.SearchResults = *triggerResults

	cycleResult.TotalSearches = triggerResults.MissingTriggered + triggerResults.CutoffTriggered
	cycleResult.TotalFailures += triggerResults.FailureCount

	// 4. Log triggered searches
	if !dryRun {
		for _, result := range triggerResults.Results {
			if result.Success {
				if len(result.ItemIDs) > 0 {
					a.logger.LogSearches(result.ServerName, result.ServerType, result.Category, len(result.ItemIDs), isManual)
				}
			} else {
				a.logger.LogSearchError(result.ServerName, result.ServerType, result.Category, result.Error)
				cycleResult.Errors = append(cycleResult.Errors, fmt.Sprintf("server %s search failed for %s (%s): %s", result.ServerName, result.Category, result.ServerType, result.Error))
			}
		}
	}

	cycleResult.Duration = time.Since(startTime)
	a.logger.LogCycleEnd(cycleResult.TotalSearches, cycleResult.TotalFailures, isManual)

	if len(cycleResult.Errors) > 0 {
		return cycleResult, fmt.Errorf("automation cycle completed with %d errors", len(cycleResult.Errors))
	}

	return cycleResult, nil
}
